### DISCUSSION TITLE
Full sequence re-estimation in each measurement

### DISCUSSION THREAD

**Discussion by adielstatman** [dstl/Stone-Soup#582] [Q&A]
Hi.
I want to demonstrate a multi-target tracking which is not "incremental", i.e. that a new measurement that is input, can affect the former estimations (the whole tracking). 
This should be operated since thanks to the new measurement, the former measurements' association probabilities are updated and this should influence, e.g.,  on the estimation weights. 

Since tutorial 8 demonstrates probabilistic data association, what should I update in the code of tutorial 8 in order to get such estimation? 
Thanks,
Adiel.

**Comment by sdhiscocks** [comment]
[Smoother](https://stonesoup.readthedocs.io/en/latest/stonesoup.smoother.html) may be of interest, but doesn't look at updating association weights.

**Comment by orosoman-dstl** [comment]
You might want to have a look at multi-frame assignment, implemented in PR #568.

**Comment by adielstatman** [comment]
> [Smoother](https://stonesoup.readthedocs.io/en/latest/stonesoup.smoother.html) may be of interest, but doesn't look at updating association weights.

Thank you.
When adding in Tutorial 8 

> from stonesoup.smoother.kalman import ExtendedKalmanSmoother as KS
> Smoother=KS(transition_model=transition_model)

and in the end of the tracks loop in the bottom cell

for track in tracks:
> ...
>  track=Smoother.smooth(track)



I get:

>   File "/home/adiel/anaconda3/lib/python3.8/site-packages/stonesoup/smoother/kalman.py", line 78, in _prediction
>     return state.hypothesis.prediction
> 
> AttributeError: 'MultipleHypothesis' object has no attribute 'prediction'

**Reply by sdhiscocks** [reply]
Looks like regression from previous refactor of smoothers, so they don't function with mixture type tracks. The below should resolve the issue.
```diff
diff --git a/stonesoup/smoother/kalman.py b/stonesoup/smoother/kalman.py
index e332ae13..f9ff2221 100644
--- a/stonesoup/smoother/kalman.py
+++ b/stonesoup/smoother/kalman.py
@@ -6,6 +6,7 @@ import numpy as np
 
 from .base import Smoother
 from ..base import Property
+from ..types.multihypothesis import MultipleHypothesis
 from ..types.prediction import Prediction, GaussianStatePrediction
 from ..types.update import Update, GaussianStateUpdate
 from ..models.base import LinearModel
@@ -75,7 +76,17 @@ class KalmanSmoother(Smoother):
         if isinstance(state, GaussianStatePrediction):
             return state
         elif isinstance(state, GaussianStateUpdate):
-            return state.hypothesis.prediction
+            if isinstance(state.hypothesis, MultipleHypothesis):
+                predictions = {hypothesis.prediction for hypothesis in state.hypothesis}
+                if len(predictions) == 1:
+                    # One predictions, this is fine to use.
+                    return predictions.pop()
+                else:
+                    # Multiple predictions, so can't process this.
+                    raise ValueError(
+                        "Track has MultipleHypotheses updates with multiple predictions.")
+            else:
+                return state.hypothesis.prediction
         else:
             raise TypeError("States must be GaussianStatePredictions or GaussianStateUpdates.")
 ```

I'll look at opening a PR with this, but you can apply the above for now to test.

**Reply by adielstatman** [reply]
Great, it works.
Thank you!
